README: Document output JSON format (#93)
open-contracting/lib-cove-ocds: README.rst

.. |Python Version| image:: https://img.shields.io/pypi/pyversions/libcoveocds.svg
   :target: https://pypi.org/project/libcoveocds/

Output JSON format
------------------

The output is an object with the following properties:

===================================== ===================== ==============
Property (key) name		      Type                  Value
===================================== ===================== ==============

> ``file_type``                         string                The type of the file supplied, e.g. ``json``

Reviewer: Can we list all the currently supported file types?

Author: Yes.

> ``version_used``                      string                The version of the OCDS schemas used, e.g. ``1.1``

Reviewer: If the version is not specified in the OCDS file, is this `1.0`, or empty?

Author: `1.0`. I'll add a note.

> ``schema_url``                        string                The url to the package schema used, e.g. ``https://standard.open-contracting.org/1.1/en/release-package-schema.json``

Reviewer: Uppercase the URL (and CSV) acronym throughout.

Reviewer: @Bjwebb

``extensions``                        object                An extensions_ object
``validation_errors``                 array[object]         An array of validation_errors_ objects

> ``common_error_types``                array[]               Always an empty array.

Reviewer: Here and elsewhere, omit the ending period, for consistency with other cells.

> ``common_error_types``                array[]               Always an empty array.

Reviewer: Why always empty?

Reviewer: @Bjwebb

Author: Because it's only used by other coves.

Reviewer: Out of curiosity, what are the semantics in the other CoVEs?

Author: It's a [list of types and formats](https://github.com/ThreeSixtyGiving/dataquality/blob/02c6a70c30413c76234773afc46fa1a095abde5c/lib360dataquality/cove/threesixtygiving.py#L355-L362) to create a "more info" link for, in the validation error section of the CoVE web interface. These then link to a common errors page https://dataquality.threesixtygiving.org/common_errors#date-time. It's necessary because not all types and formats are on that page, so we don't want to make useless links.

Reviewer: Aha. In the OCDS DRT we would prefer to have such instructions on the results page rather than a separate page, so OK to leave this as just "Always an empty array".

``deprecated_fields``                 array[object]         An array of deprecated_fields_ objects
``releases_aggregates``               object                A releases_aggregates_ object
``records_aggregates``                object                A records_aggregates_ object
``additional_closed_codelist_values`` object                A mapping from from codelist path to an `additional codelist object`_.

> ``additional_open_codelist_values``   object                A mapping from from codelist path to an `additional codelist object`_.

Reviewer: What is a codelist path? `codelists/name.csv` ? `name.csv`? `name`?

```suggestion
``additional_closed_codelist_values`` object                A mapping from a codelist path to an `additional codelist object`_
``additional_open_codelist_values``   object                A mapping from a codelist path to an `additional codelist object`_
```

> ``additional_closed_codelist_values`` object                A mapping from a codelist JSON Pointer (with array indices removed, e.g. ``releases/tender/documents/documentType``), to an `additional codelist object`_
> ``additional_open_codelist_values``   object                A mapping from a codelist JSON Pointer (with array indices removed, e.g. ``releases/tender/documents/documentType``), to an `additional codelist object`_

Reviewer: ```suggestion
``additional_closed_codelist_values`` object                A mapping from a codelist field's JSON Pointer (with array indices removed, e.g. ``releases/tender/documents/documentType``) to an `additional codelist object`_
``additional_open_codelist_values``   object                A mapping from a codelist field's JSON Pointer (with array indices removed, e.g. ``releases/tender/documents/documentType``) to an `additional codelist object`_
```

``additional_checks``                 object                A mapping from an additional check type (currently only ``empty_field``) to an array of `additional check objects <additional check object_>`_

> ``conformance_errors``                object                See conformance_errors_

Reviewer: ```suggestion
``conformance_errors``                object                A conformance_errors_ object
```

> ``additional_fields``                 array[object]         Just the top level additional fields, see additional_fields_

Reviewer: ```suggestion
``additional_fields``                 array[object]         The top-level additional fields, as an array of additional_fields_ objects
```

> ``all_additional_fields``             array[object]         All additional fields, including children of other additional fields, see all_additional_fields_

Reviewer: ```suggestion
``all_additional_fields``             array[object]         All additional fields, including children of other additional fields, as an array of all_additional_fields_ objects
```

> ``ocds_prefixes_bad_format``          array[]               Always an empty array. This is `a bug <https://github.com/open-contracting/lib-cove-ocds/issues/94>`_, see conformance_errors_ for where this property is actually populated.

Reviewer: ```suggestion
``ocds_prefixes_bad_format``          array[]               Always an empty array. This is `a bug <https://github.com/open-contracting/lib-cove-ocds/issues/94>`_. See the ``ocds_prefixes_bad_format`` property in the conformance_errors_ object instead.
```

===================================== ===================== ==============

Note that wherever a schema is used, it is the extended schema (if extensions exist).

extensions
^^^^^^^^^^

============================= ===================== ==============
Property (key) name	      Type                  Value
============================= ===================== ==============
``extensions``                array[object]         An `extensions/extensions`_ object

> ``invalid_extensions``        array[array[string]]  An array of pairs representing the extension url, and a human readable error message, e.g. ``[["http://etc", "404: not found"]]``

Reviewer: ```suggestion
``invalid_extensions``        array[array[string]]  An array of pairs of an extension URL and a human-readable error message, e.g. ``[["http://etc", "404: not found"]]``
```

``extended_schema_url``       string                The file the extended schema will be written to, if an output directory has been set, e.g. ``extended_schema.json``           
``is_extended_schema``        boolean               Has the schema been extended?
============================= ===================== ==============

extensions/extensions
^^^^^^^^^^^^^^^^^^^^^

======================= =============== ============
Property (key) name     Type            Value
======================= =============== ============

> ``url``                 string          The url of the extension's extension.json, e.g. ``https://raw.githubusercontent.com/open-contracting-extensions/ocds_metrics_extension/master/extension.json``

Reviewer: ```suggestion
``url``                 string          The URL of the extension's metadata file, e.g. ``https://raw.githubusercontent.com/open-contracting-extensions/ocds_metrics_extension/master/extension.json``
```

> ``schema_url``          string          The url of the extension's schema json e.g. ``https://raw.githubusercontent.com/open-contracting-extensions/ocds_metrics_extension/master/release-schema.json``

Reviewer: I assume this is for the release schema file only.

```suggestion
``schema_url``          string          The URL of the extension's release schema file, e.g. ``https://raw.githubusercontent.com/open-contracting-extensions/ocds_metrics_extension/master/release-schema.json``
```

Reviewer: @Bjwebb

Author: Yes, this is always the release schema https://github.com/open-contracting/lib-cove-ocds/blob/ef9d52bf375ff09e59fcd8cb23860adeac63328e/libcoveocds/schema.py#L221

``description``         string          Taken from the extension.json
``name``                string          Taken from the extension.json
``documentationUrl``    string          Taken from the extension.json
``failed_codelists``    object          A mapping from extended codelist name (prefixed with ``+`` or ``-`` if appropriate) to human readable strings describing the error.

> ``codelists``           array[string]   Taken from the extension.json

Reviewer: ```suggestion
``description``         string          Extracted from the metadata file
``name``                string          Extracted from the metadata file
``documentationUrl``    string          Extracted from the metadata file
``failed_codelists``    object          A mapping from an extended codelist name (prefixed with ``+`` or ``-`` if appropriate) to a human-readable error message
``codelists``           array[string]   Extracted from the metadata file
```

======================= =============== ============

validation_errors
^^^^^^^^^^^^^^^^^

Note that this list will exclude codelist errors, which instead appear in ``additional_closed_codelist_values``.

======================= =========== ========
Property (key) name     Type        Value
======================= =========== ========

> ``type``                string      The JSON schema keyword that caused the validation error, e.g. ``minLength`` (`full list in the jsonschema lib <https://github.com/Julian/jsonschema/blob/9b6a9f5a4b7341cdbfc3cbee32d66bc190e4ced8/jsonschema/validators.py#L321-L345>`_), unless the schema keyword was ``type`` or ``format``, in which case this value is the relevant `type <https://datatracker.ietf.org/doc/html/draft-zyp-json-schema-04#section-3.5>`_ or `format <https://datatracker.ietf.org/doc/html/draft-fge-json-schema-validation-00#section-7.3>`_) e.g. ``array`` or ``date-time``

Reviewer: ```suggestion
``type``                string      The JSON Schema keyword that caused the validation error, e.g. ``minLength`` (`full list in the jsonschema lib <https://github.com/Julian/jsonschema/blob/9b6a9f5a4b7341cdbfc3cbee32d66bc190e4ced8/jsonschema/validators.py#L321-L345>`_), unless the keyword is ``type`` or ``format``, in which case this is the relevant `type <https://datatracker.ietf.org/doc/html/draft-zyp-json-schema-04#section-3.5>`_ or `format <https://datatracker.ietf.org/doc/html/draft-fge-json-schema-validation-00#section-7.3>`_, e.g. ``array`` or ``date-time``
```

> ``field``               string      Like ``path``, but with array indices removed e.g. ``releases/tender/items``

Reviewer: ```suggestion
``field``               string      The JSON Pointer to the erroneous data, with array indices removed, e.g. ``releases/tender/items``
```

> ``description``         string      A human readable message about the error, e.g. ``'id' is missing but required within 'items'``

Reviewer: ```suggestion
``description``         string      A human-readable error message, e.g. ``'id' is missing but required within 'items'``
```

> ``path``                string      The JSON Pointer to the erroneous data e.g. ``releases/0/tender/items/0``

Reviewer: Put a comma before "e.g.", in all cases.

Reviewer: @Bjwebb

> ``value``               any         The value in the data that was erroneous, or `""` if not applicable.

Reviewer: ```suggestion
``value``               any         The value in the data that was erroneous, or ``""`` if not applicable
```

======================= =========== ========

deprecated_fields
^^^^^^^^^^^^^^^^^

======================================= =========================== ==============
Property (key) name	                Type                        Value
======================================= =========================== ==============

> ``paths``                               array[string]               An array of JSON Pointers to parent objects with some deprecated data e.g. ``["releases/0/tender"]``

Reviewer: ```suggestion
``paths``                               array[string]               An array of JSON Pointers to parent objects containing deprecated fields, e.g. ``["releases/0/tender"]``
```

> ``explanation``                         array[string]               A pair of deprecated version, and a human readable explanation., e.g. ``["1.1", "Some explanation text"]``

Reviewer: ```suggestion
``explanation``                         array[string]               A pair of the version in which the field was deprecated, and the human-readable deprecation message, e.g. ``["1.1", "Some explanation text"]``
```

> ``field``                               string                      The field within the parent object that is deprecated, e.g. ``amendment``

Reviewer: ```suggestion
``field``                               string                      The name of the field within the parent object that is deprecated, e.g. ``amendment``
```

``implementation_doctype``              object                      A mapping from ``documentType``, to the number of occurrences.
``implementation_milestones_doctype``   object                      A mapping from ``documentType``, to the number of occurrences.
``contracts_without_awards``            array                       An array of contract objects (from the data) that don't have awards.
======================================= =========================== ==============

records_aggregates
^^^^^^^^^^^^^^^^^^

============================= ==================== ==============
Property (key) name	      Type                 Value
============================= ==================== ==============
``count``                     integer              The number of items in the records array
``unique_ocids``              array*               An array of all ocids, deduplicated.
============================= ==================== ==============

additional codelist object
^^^^^^^^^^^^^^^^^^^^^^^^^^

=========================== ======================= ============
Property (key) name	    Type                    Value
=========================== ======================= ============

> ``path``                    string                  The path of the parent object, e.g. ``releases/tender/documents``

Reviewer: I assume array indices are removed.

```suggestion
``path``                    string                  The JSON Pointer to the parent object, with array indices removed, e.g. ``releases/tender/documents``
```

> ``field``                   string                  The JSON property name, e.g. ``documentType`` 

Reviewer: ```suggestion
``field``                   string                  The name of the codelist field, e.g. ``documentType`` 
```

> ``codelist``                string                  The csv file containing the codelist, e.g. ``documentType.csv``

Reviewer: ```suggestion
``codelist``                string                  The filename of the codelist, e.g. ``documentType.csv``
```

> ``codelist_url``            string                  A url that the codelist csv is accessible at, e.g. ``https://raw.githubusercontent.com/open-contracting/standard/1.1/schema/codelists/documentType.csv``

Reviewer: ```suggestion
``codelist_url``            string                  The URL of the codelist, e.g. ``https://raw.githubusercontent.com/open-contracting/standard/1.1/schema/codelists/documentType.csv``
```

> ``codelist_amend_urls``     array[array[string]     Urls to codelist csvs from extensions that amend the codelist. Structure is an array of pairs of ``+`` or ``-`` and the url. e.g. ``[["+", "https://raw.githubusercontent.com/open-contracting-extensions/ocds_tariffs_extension/d9df2969030b0a555c24c7db685262c714b4da24/codelists/+documentType.csv"]]``

Reviewer: ```suggestion
``codelist_amend_urls``     array[array[string]     The URLs of the codelist patches in extensions that modify the codelist, as an array of pairs of ``+`` or ``-`` and the URL, e.g. ``[["+", "https://raw.githubusercontent.com/open-contracting-extensions/ocds_tariffs_extension/d9df2969030b0a555c24c7db685262c714b4da24/codelists/+documentType.csv"]]``
```

``isopen``                  boolean                 Is this an open codelist?

> ``values``                  array*                  Values in the data that were not on the codelist

Reviewer: ```suggestion
``values``                  array*                  Values of the field that are not in the codelist
```

> ``extension_codelist``      boolean                 Was the codelist added by an extension? (Not just modified by it).

Reviewer: ```suggestion
``extension_codelist``      boolean                 Is the codelist added by an extension? (Not only modified by it)
```

=========================== ======================= ============

additional check object
^^^^^^^^^^^^^^^^^^^^^^^

=========================== ===================== ==============
Property (key) name	    Type                  Value
=========================== ===================== ==============

> ``json_location``           string                A JSON Pointer to the data that was problematic e.g. ``releases/0/buyer``

Reviewer: ```suggestion
``json_location``           string                A JSON Pointer to the problematic data, e.g. ``releases/0/buyer``
```

=========================== ===================== ==============

conformance_errors
^^^^^^^^^^^^^^^^^^

=============================== ======================= =====
Property (key) name	        Type                    Value
=============================== ======================= =====

> ``ocds_prefixes_bad_format``    array[array[string]]    An array of pairs of the bad ocid value, and the JSON Pointer to it, e.g. ``["MY-ID", "releases/0/ocid"]``

Reviewer: ```suggestion
``ocds_prefixes_bad_format``    array[array[string]]    An array of pairs of a bad ``ocid`` value and the JSON Pointer to it, e.g. ``["MY-ID", "releases/0/ocid"]``
```

> ``ocid_description``            string                  The descriptive text about ocids taken from the schema.

Reviewer: ```suggestion
``ocid_description``            string                  The description of the ``ocid`` field from the OCDS schema
```

Reviewer: @Bjwebb

> ``ocid_info_url``               string                  The url to the information about identifiers in the reference docs

Reviewer: ```suggestion
``ocid_info_url``               string                  The URL to the identifiers content in the OCDS documentation
```

=============================== ======================= =====

additional_fields
^^^^^^^^^^^^^^^^^

============================= ========= ==============
Property (key) name	      Type      Value
============================= ========= ==============

> ``path``                      string    The path of the parent object, e.g. ``/publisher``

Reviewer: I don't know if array indices are removed or not.

```suggestion
``path``                      string    The JSON Pointer to the parent object, e.g. ``/publisher``
```

> ``field``                     string    The JSON property name, e.g. ``myField``

Reviewer: ```suggestion
``field``                     string    The name of the additional field, e.g. ``myField``
```

> ``usage_count``               integer   How many times this additional field appears

Reviewer: ```suggestion
``usage_count``               integer   The number of times the additional field is set
```

============================= ========= ==============

all_additional_fields
^^^^^^^^^^^^^^^^^^^^^

=================================== =========== ==============
Property (key) name	            Type        Value
=================================== =========== ==============

> ``count``                           integer     How many times this additional field appears

Reviewer: ```suggestion
``count``                           integer     The number of times the additional field is set
```

> ``examples``                        array*      An array of values for this field

Reviewer: What is the maximum length of this array?

```suggestion
``examples``                        array*      A sample of values of the field
```

Author: 3. I'll add this.

> ``root_additional_field``           boolean     Is this the first additional field we find descending into this bit of the schema? ie. is the parent in the schema?

Reviewer: ```suggestion
``root_additional_field``           boolean     Is the parent object described by the schema?
```

> ``additional_field_descendance``    object      Only appears if ``root_additional_field`` is true. A mapping from paths, to objects like those in all_additional_fields_, for each of the additional fields that can be found by descending into the data from this field.

Reviewer: Isn't this description circular? ("like those in all_additional_fields_")

```suggestion
``additional_field_descendance``    object      Is only set if ``root_additional_field`` is true. A mapping from paths to objects, like those in all_additional_fields_, for each of the additional fields that can be found by descending into the data from this field.
```

Author: It's a recursive structure, except that `root_additional_field` is always false on the second level, so it doesn't get any further than that.

Reviewer: Thanks, I've suggested new text in a new review.

> ``path``                            string      The path of the parent object, e.g. ``/publisher``

Reviewer: I don't know if array indices are removed or not.

```suggestion
``path``                            string      The JSON Pointer to the parent object, e.g. ``/publisher``
```

> ``field_name``                      string      The JSON property name, e.g. ``myField``

Reviewer: ```suggestion
``field_name``                      string      The name of the additional field, e.g. ``myField``
```

> ``additional_field_descendance``    object      Is only set if ``root_additional_field`` is true. A mapping from JSON Pointers without array indices to objects, like those in all_additional_fields_, for each of the additional fields that can be found by descending into the data from this field

Reviewer: ```suggestion
``additional_field_descendance``    object      The additional fields that are descendants of this field. Is only set if ``root_additional_field`` is true. A mapping from an additional field's JSON Pointer (with array indices removed) to an all_additional_fields_ object in which ``root_additional_field`` is false
```
